v3/MmlTree/ts/Node.ts

```diff
@@ -0,0 +1,355 @@
+     */
+    getProperty(name: string): Property;
+
+    /*
+     * @return {string[]}  An array of the names of every property currently defined
+     */
+    getPropertyNames(): string[];
+
+    /*
+     * @return {PropertyList}  The propery list containing all the properties of the node
+     */
+    getAllProperties(): PropertyList;
+
+    /*
+     * @param {string[]} names  The names of the properties to be removed
+     */
+    removeProperty(...names: string[]): void;
+
+
+    /*
+     * @param {string} kind  The type of node to test for
+     * @return {boolean}  True when the node is of the given type
+     */
+    isKind(kind: string): boolean;
+
+    /*
+     * @param {Node[]} children  The child nodes to add to this node
+     */
+    setChildren(children: Node[]): void;
+
+    /*
+     * @param {Node} child  A node to add to this ones children
+     * @return {Node}  The child node that was added
+     */
+    appendChild(child: Node): Node;
+
+    /*
+     * @param {Node} newChild  A child node to be replaced
+     * @param {Node} oldChild  A child node to be replaced
+     * @return {Node}  The old child node that was removed
+     */
+    replaceChild(newChild: Node, oldChild: Node): Node;
+
+    /*
+     * @param {Node} child  A child node whose index in childNodes is desired
+     * @return {number}  The index of the child in childNodes, or null if not found
+     */
+    childIndex(child: Node): number;
+
+    /*
+     * @param {string} kind  The kind of nodes to be located in the tree
+     * @return {Node[]}  An array of nodes that are children (at any depth) of the given kind
+     */
+    findNodes(kind: string): Node[];
+
+    /*
+     * @param {Function} func  A function to apply to each node in the tree rooted at this node
+     * @param {any} data  data to pass to the function (as state information)
+     */
+    walkTree(func: (node: Node, data?: any) => void, data?: any): void;
+}
+
+/*********************************************************/
+/*
+ *  The generic Node class interface
+ */
+
+export interface INodeClass {
+    /*
+     * @param {NodeFactory} factory  The NodeFactory to use to create new nodes when needed
+     * @param {PropertyList} properties  Any properties to be added to the node, if any
+     * @param {Node[]} children  The initial child nodes, if any
+     * @return {Node}  The newly created node
+     */
+    new (factory: NodeFactory, properties?: PropertyList, children?: Node[]): Node;
+}
+
+/*********************************************************/
+/*
+ *  The abstract Node class
+ */
+
+export abstract class ANode implements INode {
+
+    /*
+     * The parent node for this one
+     */
+    public parent: Node = null;
+
+    /*
+     * The properties for this node
+     */
+    protected properties: PropertyList = {};
+
+    /*
+     * The NodeFactory to use to create additional nodes, as needed
+     */
+    public factory: NodeFactory = null;
```

Comment:
It didn't seem necessary to add that overhead in order to prevent something we weren't going to do anyway.